tests/integration/utils.sh

```diff
@@ -0,0 +1,132 @@
+  local working_directory
+  local output # output of podman used to check if containers exist or not
+
+  for distro in "${DISTROS[@]}"; do
+    # Only build the image if it does not exists. That's because trying to build
+    # the podman image takes a half second even if it exists and is cached.
+    if ! podman image exists "kw-${distro}"; then
+      # Build the image or exit. The integration tests cannot continue otherwise
+      if ! build_distro_image "$distro" > /dev/null; then
+        complain 'failed to setup container environment'
+        teardown_container_environment
+        exit 1
+      fi
+    fi
+
+    # the name of the container and the container image.
+    container_img="kw-${distro}"
+    container_name="kw-${distro}"
+
+    # Get running containers matching the container name.
+    # The output of the command will be a list of container names, one per line
+    output=$(podman container list --filter name="${container_name}" --format '{{.Names}}')
+
+    # If container is running, we do not recreate it for optimization purposes
+    # unless it is outdated.
+    if [[ "${output}" == "${container_name}" ]]; then
+      # to check if it is outdated, we check kw's version refers to the same
+      # commit as this local kw repo.
+
+      # first, get the sha inside the container
+      sha_container=$(podman exec "${container_name}" kw --version | grep Commit | sed 's/Commit: //')
+
+      # get the sha of this repo
+      sha_kw=$(git --git-dir "${KWROOT_DIR}/.git" rev-parse --short HEAD)
+
+      # if container is up to date, we will reuse it.
+      # Otherwise, destroy the container and create a new one.
+      if [[ "${sha_kw}" == "${sha_container}" ]]; then
+        continue
+      else
+        teardown_single_container "${container_name}"
+      fi
+    fi
+
+    # Maybe the container is not running, but it does exist.
+    # In this case, we tear it down because it could be in a improper state.
+    if podman container exists "${container_name}"; then
+      teardown_single_container "${container_name}"
+    fi
+
+    # containers are isolated environments designed to run a  process.  After  the
+    # process ends, the container is destroyed. In order execute multiple commands
+    # in the container, we need to keep the container,  which  means  the  primary
+    # process must not terminate. Therefore, we run a never-ending command as  the
+    # primary process,  so  that  we  can  execute  multiple  commands  (secondary
+    # processes) and get the output of each of them separately.
+    working_directory=/tmp/kw
+    podman run \
+      --workdir "${working_directory}" \
+      --volume "${KWROOT_DIR}":"${working_directory}" \
+      --env PATH='/root/.local/bin:/usr/bin' \
+      --name "${container_name}" \
+      --detach \
+      "${container_img}" sleep infinity > /dev/null
+
+    # install kw again
+    podman exec "${container_name}" \
+      ./setup.sh -i --force --skip-docs > /dev/null 2>&1
```

Comment:
Same thing, check the result. Also, if this command fails, it means one of the containers cannot be used, right? If that is true, you probably need a mechanism to deal with that case, otherwise, some tests will fail.